Add Session workflows documentation (#808)
jupyter-server/jupyter_server: docs/source/developers/architecture.rst

   request has the Session ID that kernel is currently using.

#. **Session Manager** gets the Session data from the SQLite3 DataBase and sends
   the kernel ID to **Mapping Kernel Manager**.

#. **Mapping Kernel Manager** starts the kernel shutdown process by using
   **Multi Kernel Manager** and **Kernel Manager**.

#. **Kernel Manager** receives interrupt mode from **Kernel Spec Manager**.
   **Kernel Manager** supports ``Signal`` and ``Message`` interrupt modes.
   By default, the ``Signal`` interrupt mode is being used.

   - When the interrupt mode is ``Signal``, the **Kernel Provisioner**
     interrupts the kernel with the ``SIGINT`` operating system signal
     (although other provisioner implementations may use a different approach).

   - When interrupt mode is ``Message``, Session sends
     the `"interrupt_request" <https://jupyter-client.readthedocs.io/en/latest/messaging.html#msging-interrupt>`_
     message on the control channel.

#. After interrupting kernel, Session sends the `"shutdown_request" <https://jupyter-client.readthedocs.io/en/latest/messaging.html#kernel-shutdown>`_
   message on the control channel.

> #. **Kernel Manager** waits for the kernel shutdown. After the timeout,

Reviewer: ```suggestion
#. **Kernel Manager** waits for the kernel shutdown. After the timeout, and if it detects the kernel process is still running, the 
```

   **Kernel Manager** interrupts the kernel with ``SIGTERM`` operating system signal.

#. **Kernel Manager** clean up the kernel resources. It removes kernel's interprocess
   communication ports, closes control socket, and releases Shell, IOPub, StdIn,
   Control, and Heartbeat ports.

#. When shutdown is finished, **Session Manager** deletes the Session data from

>    the SQLite3 DataBase and responses 204 Status Code to the Notebook client.

Reviewer: ```suggestion
   the SQLite3 database and responses 204 status code to the Notebook client.
```

    and provides a list of available kernel configurations. To learn about
    Kernel Spec Manager, check `the Jupyter Client guide <https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs>`_.

Create Session Workflow
-----------------------

The create Session workflow can be seen in the figure below:

.. image:: ../images/session-create.drawio.png
   :alt: Create Session Workflow
   :width: 90%
   :align: center

When a user starts a new kernel, the following steps occur:

#. The Notebook client sends |create_session|_ request to Jupyter Server. This
   request has all necessary data, such as Notebook name, type, path, and kernel
   name.

#. **Session Manager** asks **Contents Manager** for the kernel file system path
   based on the input data.

#. **Session Manager** sends kernel path to **Mapping Kernel Manager**.

#. **Mapping Kernel Manager** starts the kernel create process by using
   **Multi Kernel Manager** and **Kernel Manager**. You can learn more about
   **Multi Kernel Manager** in
   `the Jupyter Client APIs <https://jupyter-client.readthedocs.io/en/latest/api/manager.html#multikernelmanager-controlling-multiple-kernels>`_.

> #. **Kernel Manager** uses provisioner layer to launch a new kernel.

Reviewer 2: ```suggestion
#. **Kernel Manager** uses the provisioner layer to launch a new kernel.
```

  - **Kernel Spec Manager** parses files with JSON specification for a Kernels,
    and provides a list of available Kernel configurations. To learn about
    Kernel Spec, check `the Jupyter Client guide <https://jupyter-client.readthedocs.io/en/stable/kernels.html#kernel-specs>`_.

Create Session Workflow
-----------------------

> The create Session workflow can be seen in figure below:

Reviewer: ```suggestion
The create Session workflow can be seen in the figure below:
```

.. image:: ../images/session-create.drawio.png
   :alt: Create Session Workflow
   :width: 90%
   :align: center

> When a user starts a new Kernel, the following steps process:

Reviewer: ```suggestion
When a user starts a new kernel, the following steps occur:
```

Reviewer: I think "Kernel" should be "kernel" except when it's a proper noun (e.g., Mapping Kernel Manager, Kernel Spec Manager, etc.).  I believe this was also true in the general architecture section as well.  Perhaps a sweep can be made for such instances.

Author: Sounds good, I will change it to "kernel".

#. Notebook client receives the created Session data.

.. _create_session: https://petstore.swagger.io/?url=https://raw.githubusercontent.com/jupyter/jupyter_server/master/jupyter_server/services/api/api.yaml#/sessions/post_api_sessions

.. |create_session| replace:: the *POST /api/sessions*

Delete Session Workflow
-----------------------

The delete Session workflow can be seen in figure below:

.. image:: ../images/session-delete.drawio.png
   :alt: Delete Session Workflow
   :width: 80%
   :align: center

When a user stops a Kernel, the following steps process:

#. The Notebook client sends |delete_session|_ request to Jupyter Server. This
   request has the Session ID that Kernel is currently using.

#. **Session Manager** gets the Session data from the SQLite3 DataBase and sends
   the Kernel ID to **Mapping Kernel Manager**.

#. **Mapping Kernel Manager** starts the Kernel shutdown process by using
   **Multi Kernel Manager** and **Kernel Manager**.

#. **Kernel Manager** receives interrupt mode from **Kernel Spec Manager**.
   **Kernel Manager** supports ``Signal`` and ``Message`` interrupt modes.
   By default, the ``Signal`` interrupt mode is being used.

>    - When interrupt mode is ``Signal``, **Kernel Provisioner** kills the Kernel
>      with ``SIGINT`` operating system signal.

Reviewer: ```suggestion
   - When the interrupt mode is ``Signal``, the **Kernel Provisioner** interrupts the kernel
     with the ``SIGINT`` operating system signal (although other provisioner implementations may use a different approach).
```

Reviewer: The diagram for the signal-based interrupt should probably replace `Kill` with `Interrupt` as well.

   - When interrupt mode is ``Message``, **Session** sends
     the ``interrupt_request`` message on the control channel. Check the
     `messaging in Jupyter guide <https://jupyter-client.readthedocs.io/en/latest/messaging.html#msging-interrupt>`_
     to know more about it.

#. When Kernel is shutdown, **Session Manager** deletes the Session data from

>    the SQLite3 DataBase and responses 204 Status Code to the Notebook client.

Reviewer: Should the kernel's shutdown be detailed similar to its interrupt?  I know that can be a lot and we might be trying to limit text.  Perhaps we could do something like...

When the [kernel is shutdown](https://github.com/jupyter/jupyter_client/blob/00987a97b0b52431287c60060c98518cf36072f1/jupyter_client/manager.py#L467-L480), ...

and point to the code (or documentation produced from the code)?

Author: @kevin-bates I think that makes sense. I added steps after kernel interrupt.
Please let me know, what do you think.
